workitem/workitem_repository.go

```diff
@@ -93,6 +114,114 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
+// Version must be the same as the one int the stored version
+func (r *GormWorkItemRepository) Reorder(ctx context.Context, beforeID string, wi app.WorkItem) (*app.WorkItem, error) {
+	var order float64
+	res := WorkItem{}
+	beforeItem := WorkItem{}
+	afterItem := WorkItem{}
+	id, err := strconv.ParseUint(wi.ID, 10, 64)
+	if err != nil || id == 0 {
+		return nil, errors.NewNotFoundError("work item", wi.ID)
+	}
+
+	tx := r.db.First(&res, id)
+	if tx.RecordNotFound() {
+		return nil, errors.NewNotFoundError("work item", wi.ID)
+	}
+	if tx.Error != nil {
+		return nil, errors.NewInternalError(err.Error())
+	}
+	if res.Version != wi.Version {
+		return nil, errors.NewVersionConflictError("version conflict")
+	}
+
+	wiType, err := r.wir.LoadTypeFromDB(ctx, wi.Type)
+	if err != nil {
+		return nil, errors.NewBadParameterError("Type", wi.Type)
+	}
+	if beforeID != "" {
+		beforeId, err := strconv.ParseUint(beforeID, 10, 64)
+		if err != nil || beforeId == 0 {
+			return nil, errors.NewNotFoundError("work item", string(beforeId))
+		}
+		tx = r.db.First(&beforeItem, beforeId)
+		if tx.RecordNotFound() {
+			return nil, errors.NewNotFoundError("work item", string(beforeId))
+		}
+		if tx.Error != nil {
+			return nil, errors.NewInternalError(err.Error())
+		}
+		beforeOrder, err := strconv.ParseFloat(fmt.Sprintf("%v", beforeItem.Executionorder), 64)
+		if err != nil {
+			return nil, errors.NewBadParameterError("data.attributes.order", res.Executionorder)
+		}
+		tx2 := r.db.Where("executionorder < ?", beforeItem.Executionorder).Order("executionorder desc", true).Last(&afterItem)
+		if afterItem.ID == 0 {
+			// The item is moved to first position
+			order = (0 + beforeOrder) / 2
+		} else {
+			if afterItem.Executionorder == res.Executionorder {
+				// when a workitem is reordered at the same position again and again
+				order = afterItem.Executionorder
+			} else {
+				afterOrder, err := strconv.ParseFloat(fmt.Sprintf("%v", afterItem.Executionorder), 64)
+				if err != nil {
+					return nil, errors.NewBadParameterError("data.attributes.order", res.Executionorder)
+				}
+				if tx2.RecordNotFound() {
+					return nil, errors.NewNotFoundError("work item", string(beforeId))
+				}
+				if tx2.Error != nil {
+					return nil, errors.NewInternalError(err.Error())
+				}
+				order = (beforeOrder + afterOrder) / 2
+			}
+		}
+	} else {
+		// the item is moved at last position
+
+		tx2 := r.db.Order("executionorder desc", true).Last(&afterItem)
+		if tx2.RecordNotFound() {
+			return nil, errors.NewNotFoundError("work item", string(afterItem.ID))
+		}
+		if tx2.Error != nil {
+			return nil, errors.NewInternalError(err.Error())
+		}
+		afterOrder, _ := strconv.ParseFloat(fmt.Sprintf("%v", afterItem.Executionorder), 64)
+		order = afterOrder + orderValue
+	}
+
+	res.Version = res.Version + 1
+	res.Type = wi.Type
+	res.Fields = Fields{}
+
+	res.Executionorder = order
+	for fieldName, fieldDef := range wiType.Fields {
```

Comment:
Why convert all the fields etc?  Shouldn't we just store the new executionorder?